### PR TITLE
Make the ToolBar in the Find/Replace Overlay accessible #1910

### LINKED ISSUE #1910: [find/replace overlay] Overlay option buttons are not accessible via tab.
This Issue Documents a review-comment from the find/replace overlay PR
which could not be addressed into the current proposal
[comment](https://github.com/eclipse-platform/eclipse.platform.ui/pull/1192#issuecomment-2098206821) by @Victko

In the current find/replace dialog, it is impossible to "tab through" to access all options.
Instead, we want to be able to access all buttons, options and input-fields by tabbing into them - as is possible in the current find/replace dialog.

I have implemented a proposed fix locally [here](https://github.com/Wittmaxi/eclipse.platform.ui/pull/6/commits/361a0bfcbdcc230ff9f25c19640f5efcb090e087)

### PR DESCRIPTION
**Pull request by Wittmaxi**
Makes the Find/Replace Overlay options accessible by tabbing through the different option buttons.
Implements a new "AccessibleToolBar" class which wraps the ToolBar and allows for being natively accessibly
using the normal traversal mechanism provided by SWT.

fixes #1910

## How to test
1) open the find/replace overlay using Ctrl+F
2) using Tab, try to navigate to every search option which is enabled
3) using enter, try enabling/disabling search options or performing search tasks

The goal is to be able to perform every action in the find/replace overlay using only the keyboard and completely without using the mouse 🐁.

## Requires
* [x] #1192

### CODE REVIEW

**Comment by Wittmaxi** [comment]
This PR is a follow up, meant to be merged after https://github.com/eclipse-platform/eclipse.platform.ui/pull/1192 is merged

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceOverlay.java [hunk 1]
```diff
@@ -432,24 +431,24 @@ private void createSearchTools() {
 		createAreaSearchButton();
 
 		@SuppressWarnings("unused")
-		ToolItem separator = new ToolItem(searchTools, SWT.SEPARATOR);
+		ToolItem separator = searchTools.createToolItem(SWT.SEPARATOR);
```
Comment (CodeLtDave) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceOverlay.java:444]: It looks to me that you could extract the repetitive logic of creating buttons into a factory, I see the pattern:

(1. new ToolItem)
2. searchTools
3. setImage()
4. setToolTipText()
5. addSelectionListener...

**Comment by Wittmaxi** [comment]
@CodeLtDave I have addressed your concerns - what do you think?

**Comment by Wittmaxi** [comment]
The random failing tests are not related and documented here:
https://github.com/eclipse-platform/eclipse.platform.ui/issues/294
https://github.com/eclipse-platform/eclipse.platform.ui/issues/1974

@fedejeanne it would be good to have a final look over this and (if you tihnk it's good) have this merged :)

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package org.eclipse.ui.texteditor;
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java:14]: Copyright header is missing

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	/**
+	 * Creates a ToolItem handled by this ToolBar and returns it. Will add a
+	 * KeyListener which will catch all "Enter"-Keypresses.
+	 *
+	 * @param styleBits the StyleBits to apply to the created ToolItem
+	 * @return a newly created ToolItem
+	 */
+	public ToolItem createToolItem(int styleBits) {
+		ToolBar parent = new ToolBar(this, SWT.FLAT | SWT.HORIZONTAL);
+		ToolItem result = new ToolItem(parent, styleBits);
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: I'm not a big fan of calling variables "result" or the likes.

```suggestion
		ToolItem toolItem = new ToolItem(parent, styleBits);
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+			@Override
+			public void keyTraversed(TraverseEvent e) {
+				if (e.keyCode == SWT.CR || e.keyCode == SWT.KEYPAD_CR) {
+					result.setSelection(!result.getSelection());
+					e.doit = false;
+				}
+			}
+
+		});
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Get rid of unnecessary empty lines

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+		toolBars.add(parent);
+		return result;
+	}
+
+	private void addToolItemTraverseListener(ToolBar parent, ToolItem result) {
+		parent.addTraverseListener(new TraverseListener() {
+
+			@Override
+			public void keyTraversed(TraverseEvent e) {
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: `TraverseListener` is a functional interface. Replace by lambda

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+				}
+			}
+
+		});
+	}
+
+	@Override
+	public void setBackground(Color color) {
+		super.setBackground(color);
+		for (ToolBar bar : toolBars) { // we need to color each toolbar separately
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Why do we need to do it? Is it a bug? Which one?
Please add some more info to the comment or delete it. It doesn't  add any value the way it is right now.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
+package org.eclipse.ui.texteditor;
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java:14]: Copyright header is missing

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.widgets.ToolItem;
+
+/**
+ * Builder for ToolItems for {@link AccessibleToolBars}.
+ *
+ * @since 3.17
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: Since 3.17 here but Since 3.18 in `AccessibleToolbar`? Odd, please check.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+ */
+public class AccessibleToolItemBuilder {
+	private AccessibleToolBar accessibleToolBar;
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: This one is mandatory in your builder.

```suggestion
	private final AccessibleToolBar accessibleToolBar;
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	private int styleBits = SWT.NONE;
+	private Image image;
+	private String toolTipText;
+	private SelectionListener selectionListener;
+
+	public AccessibleToolItemBuilder(AccessibleToolBar accessibleToolBar) {
+	        this.accessibleToolBar = accessibleToolBar;
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]:
```suggestion
	        this.accessibleToolBar = Objects.requireNonNull(accessibleToolBar);
```

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.widgets.ToolItem;
+
+/**
+ * Builder for ToolItems for {@link AccessibleToolBars}.
+ *
+ * @since 3.17
+ */
+public class AccessibleToolItemBuilder {
```
Comment (fedejeanne) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: Reformat the code, I see some indentation problems. Please also reformat other code you added, just in case. Just the **added** code please.

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceOverlayTest.java [hunk 12]
```diff
@@ -34,8 +34,8 @@ public class FindReplaceOverlayTest extends FindReplaceUITest<OverlayAccess> {
 
 	@Override
 	public OverlayAccess openUIFromTextViewer(TextViewer viewer) {
-		OverlayAccess ret;
 
+		OverlayAccess ret;
```
Comment (fedejeanne) [inline tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceOverlayTest.java]: Revert this change, it's a NO-OP

**Review by fedejeanne — DISMISSED** [review]
@Wittmaxi thank you for contributing this.

I've added my comments based solely on the code, I haven't tested this yet. Once these comments are addressed, I can move to testing the functionality. Regarding that: a brief "How to test" section would help and it would also give me an idea of which cases have been tested (and which **haven't** ;-) )

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 3 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: I don't believe this improves the code quality, but I don't mind performing this change.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 6 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: I don't believe this comment is too important - in short: the ToolBar ToolItems don't all (easily) inherit the color of their parent. In particular, the "separator" is a culprit!

grafik
(screenshot after removing the for-loop above)

Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: I will keep the comment so that nobody accidentally removes the code (which might seem redundant as usually properties are passed down regularly to their children) in the future

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 8 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: Both should be 17 which is the current version!

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 10 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: I did not know you could do this! Thanks :)

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 11 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: I have performed Ctrl+Shift+F on the newly added files again. Note that sometimes the github formatting does not reflect the formatting you get inside of eclipse.

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 2 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java:14]: addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 7 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java:14]: addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 8 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: I have set both to be in version 3.17 in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 11 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: Can you please check if https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266 fixes your issues?

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 3 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: changed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 4 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Done in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 5 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Done in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 6 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Edited in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266
Note that the formatting problems that are displayed on gitub are not relevant as the formatting looks good in Eclipse:

grafik

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 9 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: fixed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 10 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: Done in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceOverlayTest.java (hunk 12 above)
Comment (Wittmaxi) [reply tests/org.eclipse.ui.workbench.texteditor.tests/src/org/eclipse/ui/workbench/texteditor/tests/FindReplaceOverlayTest.java]: Removed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/0c0d22bc8f0ea13eb6e1124c89747f7a4f02e266

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceOverlay.java (hunk 1 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/FindReplaceOverlay.java:444]: Addressed in https://github.com/eclipse-platform/eclipse.platform.ui/commit/292d692d7164d2767167f1ac58161f25dd30825e

**Comment by BeckerWdf** [comment]
Also fixes #1983

**Comment by Victko** [comment]
I reviewed the last corrections. It works perfect, exactly as it was expected.

**Comment by Wittmaxi** [comment]
I have fixed the merge issues

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * This class wraps the ToolBar to make it possible to use tabulator-keys to
+ * navigate between the buttons of a ToolBar. For this, we simulate a singular
+ * ToolBar by putting each ToolItem into it's own ToolBar and composing them
+ * into a Composite. Since the "Enter" keypress could not previously trigger
+ * activation behavior, we listen for it manually and send according events if
+ * necessary.
+ *
+ * @since 3.17
```
Comment (HeikoKlare) [inline bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Since this class is not part of the API (as it is package protected), a since tag should not be required, should it? So please remove if possible.

**Review by HeikoKlare — COMMENTED** [review]
Thank you, @Wittmaxi! The changes look good to me and work as expected. If I am not mistaken, there is an unncessary `@since` tag. Could you please remove that.

@fedejeanne have your concerns been fully addressed or are there still requested changes?

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java (hunk 13 above)
Comment (Wittmaxi) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolBar.java]: Yes, I can remove this, we are not adding any new API.

https://github.com/eclipse-platform/eclipse.platform.ui/commit/0676bfe83ef0e491dfbdd69ffcb617f0957b5df0

**Review by HeikoKlare — APPROVED** [review]

**Comment by HeikoKlare** [comment]
@Wittmaxi can you please rebase on master and resolve the conflict introduced by merging another find/replace-related PR?

File: bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java (hunk 11 above)
Comment (fedejeanne) [reply bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/texteditor/AccessibleToolItemBuilder.java]: It does, thank you

**Review by fedejeanne — COMMENTED** [review]
The code looks good now but I tested it (I had to resolve the conflicts myself) and I noticed that one can **not** navigate all options by pressing `Tab`, only 3 of them:

search-press-tab

I thought the goal of this PR was to be able to navigate also the options on the right? 

image

**Comment by HeikoKlare** [comment]
@fedejeanne Are you sure that you are testing the code of this PR? Your recording shows the existing behavior. With this PR it should behave like this and, at least, does so on my system:
find_replace_accessibility

**Comment by fedejeanne** [comment]
> @fedejeanne Are you sure that you are testing the code of this PR?

Silly me: I checked out the code, started the application and then tested in the IDE 🙄 

It looks good!

search-press-tab-OK

**Comment by HeikoKlare** [comment]
I've successfully validated that this works as expected on Windows 11, MacOS 14.5 and Ubuntu 22.04.4/X11.
